retrieval/discovery/ec2.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/log"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	ec2NameLabel = model.MetaLabelPrefix + "ec2_instance_id"
+)
+
+// EC2Discovery periodically performs EC2-SD requests. It implements
+// the TargetProvider interface.
+type EC2Discovery struct {
+	aws            *aws.Config
+	tags           model.LabelSet
+	regexTags      map[string]*regexp.Regexp
+	done           chan struct{}
+	interval       time.Duration
+	port           int
+	public         bool
+	instanceFilter []*ec2.Filter
+}
+
+// NewEC2Discovery returns a new EC2Discovery which periodically refreshes its targets.
+func NewEC2Discovery(conf *config.EC2SDConfig) *EC2Discovery {
+	creds := credentials.NewStaticCredentials(conf.Access, conf.Secret, "")
+	if conf.Access == "" && conf.Secret == "" {
+		creds = credentials.NewEnvCredentials()
+	}
+	tags := make(model.LabelSet)
+	filter := &ec2.Filter{
+		Name: aws.String("tag-key"),
+	}
+	for k, v := range conf.Tags {
+		tags[model.LabelName(k)] = model.LabelValue(v)
+		filter.Values = append(filter.Values, &k)
+	}
+	return &EC2Discovery{
+		aws: &aws.Config{
+			Region:      &conf.Region,
+			Credentials: creds,
+		},
+		tags:      tags,
+		regexTags: conf.RegexTags,
+		done:      make(chan struct{}),
+		interval:  time.Duration(conf.RefreshInterval),
+		port:      conf.Port,
+		public:    conf.Public,
+		instanceFilter: []*ec2.Filter{
+			{
+				Name:   aws.String("instance-state-name"),
+				Values: []*string{aws.String("running")},
+			},
+			filter,
+		},
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (ed *EC2Discovery) Run(ch chan<- *config.TargetGroup, done <-chan struct{}) {
+	defer close(ch)
+
+	ticker := time.NewTicker(ed.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	if err := ed.refresh(ch); err != nil {
+		log.Error(err)
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := ed.refresh(ch); err != nil {
+				log.Error(err)
+			}
+		case <-done:
+			return
+		}
+	}
+}
+
+// Sources implements the TargetProvider interface.
+func (ed *EC2Discovery) Sources() []string {
+	return []string{*ed.aws.Region}
+}
+
+func (ed *EC2Discovery) refresh(ch chan<- *config.TargetGroup) error {
+	EC2 := ec2.New(ed.aws)
+	dii := ec2.DescribeInstancesInput{
+		Filters: ed.instanceFilter,
+	}
+	tg := &config.TargetGroup{
+		Source: *ed.aws.Region,
+	}
+	if err := EC2.DescribeInstancesPages(&dii, func(p *ec2.DescribeInstancesOutput, lastPage bool) bool {
+		for _, r := range p.Reservations {
+			for _, inst := range r.Instances {
+				itags := make(map[string]string)
+				for _, t := range inst.Tags {
+					itags[*t.Key] = *t.Value
+				}
+				missing := false
+				labels := make(model.LabelSet)
+				for k, r := range ed.regexTags {
+					v, ok := itags[k]
+					if !ok || !r.MatchString(v) {
+						missing = true
+						break
+					}
+					labels[model.LabelName(k)] = model.LabelValue(v)
```

Comment:
tags shouldn't automatically become labels, the user should have to configure that with relabelling.